### DISCUSSION TITLE
Link that contains .html is broken. I have an url /hello.html/, but when I build the website it renders as /hello.html

### DISCUSSION THREAD

**Discussion by stevemarksd** [vercel/next.js#35069] [Help]
Link that contains .html is broken. I have an url /hello.html/, but when I build the website it renders as /hello.html

Do you know how to avoid removing the end slash from the url?

**Comment by icyJoseph** [comment]
Hi,

From your description is hard to follow what's going on, but would it help to set `trailingSlash: true` on `next.config.js`?

**Reply by stevemarksd** [reply]
Hi

Yes, of course is set to true.

So basically a `<Link href="/hello.html/">test</Link>` is converted to `<a href="/hello.html">test</a>` instead of  `<a href="/hello.html/">test</a>`

I have an article at `/hello.html/`, but not at `/hello.html`, so it breaks.

**Reply by icyJoseph** [reply]
There's a couple of things I don't understand here.

1. Is `hello.html` part of this project?
2. Is it setup as `pages/hello.html/index.js` or `ts`?

I've tested this tree:

```
├── next.config.js
├── package.json
├── pages
│   ├── api
│   │   └── hello.js
│   ├── _app.js
│   ├── hello.html
│   │   └── index.js
│   └── index.js
├── public
│   ├── favicon.ico
│   ├── nested.html
│   │   └── index.html
│   ├── page.html
│   └── vercel.svg
├── README.md
└── yarn.lock
```

Using this as `pages/index.js`:

```jsx
import Link from "next/link";

export default function Home() {
  return (
    <>
      <Link href="/hello.html/">
        <a>go to hello</a>
      </Link>
      <br />
      <Link href="/page.html/">
        <a>go to page</a>
      </Link>
      <br />
      <Link href="/nested.html/">
        <a>go to nested</a>
      </Link>
    </>
  );
}
```

And this as `next.config.js`:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  trailingSlash: true,
  rewrites: async () => {
    return {
      beforeFiles: [
        {
          source: "/hello.html/",
          destination: "/hello.html"
        },
        {
          source: "/nested.html/",
          destination: "/nested.html/index.html"
        },
        {
          source: "/nested.html",
          destination: "/nested.html/index.html"
        }
      ]
    };
  }
};

module.exports = nextConfig;
```

And I can arrive to any of the links listed works, as well as navigating directly to `http://localhost:3000/hello.html/`, `http://localhost:3000/page.html/`, and `http://localhost:3000/nested.html/`.

I think we need to see more of your code to make an assessment. 

3. If you are linking to a full `html` page, then why use `Link`? 

Hopefully we can sort this out, but yeah a little annoying that `href="/hello/"` renders to `href="/hello/`, but adding the `.html` drops the trailing slash.

**Reply by stevemarksd** [reply]
Hi

Thanks for the answer. Sorry, my assumption is that people usualy don't hardcode paths for generic posts, I clarified the project below.

I cannot use the rewrite unfortunately because I dont know the paths at build time.

> Is hello.html part of this project?

The page is dynamically rendered. So I have a catch all route.

Is a dynamic path in the project.

[...slug].ts

Nothing is static. I cannot use your rewrite rule because the slug is a dynamic value and people might add any kind of random value. I know is not great, but old articles we migrate from legacy system have .html in the slug.

> Is it setup as pages/hello.html/index.js or ts?

See above.

> If you are linking to a full html page, then why use Link?

Because is generated from next.js.

**Reply by stevemarksd** [reply]
Unfortunately seems like a no fix from next.js team

https://github.com/vercel/next.js/issues/18164

https://github.com/vercel/next.js/issues/16617

**Reply by icyJoseph** [reply]
Yeah I see your issue now, and I guess there's a business requirement to keep pages at `page.html` - I imagine bookmarked, and other types of systems pointing at these documents must be kept.

I just commented in another discussion about the possibility to map `/a/b/c` into `/d`, while keeping the browser at `/a/b/c`, and using `pages/d.ts` to render the content. 

```js
const rewrites = [
    {
        source: /(?<yyyy>\d{4})\/(?<mm>\d{1,2})\/(?<dd>\d{1,2})\/(?<uniqueId>(.+))\//,
        destination: '/resource/',
    }
];
```

Redirecting to the above, makes `query.yyyy`, `query.uniqueId` available at `getServerSideProps`, etc, you know those parts.

**Reply by stevemarksd** [reply]
Are you able to catch page.html at the rewrite level? I was not able to catch because was automatically skipped, let me try again later.
